primitives/src/lib.rs

```diff
@@ -296,3 +313,180 @@ pub struct CustomMetadata {
+		CurrencyId::Native
+	}
+}
+
+impl TryFrom<(&str, AssetIssuer)> for CurrencyId {
+	type Error = &'static str;
+
+	fn try_from(value: (&str, AssetIssuer)) -> Result<Self, Self::Error> {
+		let slice = value.0;
+		let issuer = value.1;
+		if slice.len() <= 4 {
+			let mut code: Bytes4 = [0; 4];
+			code[..slice.len()].copy_from_slice(slice.as_bytes());
+			Ok(CurrencyId::AlphaNum4 { code, issuer })
+		} else if slice.len() > 4 && slice.len() <= 12 {
+			let mut code: Bytes12 = [0; 12];
+			code[..slice.len()].copy_from_slice(slice.as_bytes());
+			Ok(CurrencyId::AlphaNum12 { code, issuer })
+		} else {
+			Err("More than 12 bytes not supported")
+		}
+	}
+}
+
+impl From<stellar::Asset> for CurrencyId {
+	fn from(asset: stellar::Asset) -> Self {
+		match asset {
+			stellar::Asset::Default(_) => CurrencyId::StellarNative,
+			stellar::Asset::AssetTypeNative => CurrencyId::StellarNative,
+			stellar::Asset::AssetTypeCreditAlphanum4(asset_alpha_num4) => CurrencyId::AlphaNum4 {
+				code: asset_alpha_num4.asset_code,
+				issuer: asset_alpha_num4.issuer.into_binary(),
+			},
+			stellar::Asset::AssetTypeCreditAlphanum12(asset_alpha_num12) =>
+				CurrencyId::AlphaNum12 {
+					code: asset_alpha_num12.asset_code,
+					issuer: asset_alpha_num12.issuer.into_binary(),
+				},
+		}
+	}
+}
+
+impl TryInto<stellar::Asset> for CurrencyId {
+	type Error = &'static str;
+
+	fn try_into(self) -> Result<stellar::Asset, Self::Error> {
+		match self {
+			Self::Token(t) => Err("Token {} not defined in the Stellar world."),
+			Self::ForeignAsset(f) => Err("Foreign Asset {} not defined in the Stellar world."),
+			Self::Native => Err("PEN token not defined in the Stellar world."),
+			Self::StellarNative => Ok(stellar::Asset::native()),
+			Self::AlphaNum4 { code, issuer } =>
+				Ok(stellar::Asset::AssetTypeCreditAlphanum4(AlphaNum4 {
+					asset_code: code,
+					issuer: PublicKey::PublicKeyTypeEd25519(issuer),
+				})),
+			Self::AlphaNum12 { code, issuer } =>
+				Ok(stellar::Asset::AssetTypeCreditAlphanum12(AlphaNum12 {
+					asset_code: code,
+					issuer: PublicKey::PublicKeyTypeEd25519(issuer),
+				})),
+		}
+	}
+}
+
+// impl fmt::Debug for CurrencyId {
+// 	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+// 		match self {
+// 			Self::Native => write!(f, "PEN"),
+// 			Self::StellarNative => write!(f, "XLM"),
+// 			Self::AlphaNum4 { code, issuer } => {
+// 				write!(
+// 					f,
+// 					"{{ code: {}, issuer: {} }}",
+// 					str::from_utf8(code).unwrap(),
+// 					str::from_utf8(
+// 						stellar::PublicKey::from_binary(*issuer).to_encoding().as_slice()
+// 					)
+// 					.unwrap()
+// 				)
+// 			},
+// 			Self::AlphaNum12 { code, issuer } => {
+// 				write!(
+// 					f,
+// 					"{{ code: {}, issuer: {} }}",
+// 					str::from_utf8(code).unwrap(),
+// 					str::from_utf8(
+// 						stellar::PublicKey::from_binary(*issuer).to_encoding().as_slice()
+// 					)
+// 					.unwrap()
+// 				)
+// 			},
+// 		}
+// 	}
+// }
```

Comment:
What about this part, wouldn't it be nice to be able to print the currency in logs?

For the `Token` you can use the functions that are added in the `create_currency_id` macro, eg
```
...
Self::Token(token_symbol) => {
				write!(f, "{:?} ({:?})", token_symbol.name(), token_symbol.symbol())
			},
			...
```

Reply:
Oh yeah. I forgot I'd commented this out temporarily until I figured out how to get it to work with the RuntimeDebug macro which interlay uses. I guess for now it's better to have it on both runtime and wasm than on neither. Uncommented and pushed the change now. Thanks! 👍